doc: vm, run http server in vm by requiring

eljefedelrodeodeljefe opened the pull request:
As user in order to run `node` code in a `vm`, you would need to apply some objects to it. This is not quite well documented and may or may not have been reason for a series of issues and SO questions.

This PR would add an example to explain this. It is to some extent controversial and was discussed in a code PR https://github.com/nodejs/node/pull/4955. Issue would have been https://github.com/nodejs/node-v0.x-archive/issues/9211 and [this link to SO](http://stackoverflow.com/questions/20899863/the-module-property-is-undefined-when-using-vm-runinthiscontext).

An easier solution would be to use `eval()`, which I regard as an anti-pattern.
~~Also it uses `require('module').wrap(code)`, which I believe to be a "private" API.~~ I have taken and modified this example from the referenced issues.

/cc @nodejs/documentation

vkurchatkin commented on doc/api/vm.markdown:
```diff
@@ -301,6 +301,33 @@ e.g. `(0,eval)('code')`. However, it also has the following additional options:
+console.log('Server running at http://127.0.0.1:8124/');
+`
+
+let wrap = require('module').wrap(code)
```
`module` is private

eljefedelrodeodeljefe replied:
Please make valuable comments that bring the discussion forth not back. Any other way then than this and eval?

vkurchatkin replied:
```js
vm.runInThisContext(`
(function(require) {
  require('http')
})
`)(require)
```

eljefedelrodeodeljefe commented:
Forced push this. Doesn't look pretty but solves the problem.

kelthenoble commented:
:+1: @eljefedelrodeodeljefe

jasnell commented on doc/api/vm.markdown:
```diff
@@ -301,6 +301,34 @@ e.g. `(0,eval)('code')`. However, it also has the following additional options:
+const vm = require('vm')
+
+let code = `
+(function(require) {
```
minor nit... having the `let code =` on a separate line with no indenting on the next line makes it a bit confusing.

eljefedelrodeodeljefe replied:
agree. fixed and force pushed. Thx.

jasnell commented:
LGTM

silverwind commented on doc/api/vm.markdown:
```diff
@@ -301,6 +301,34 @@ e.g. `(0,eval)('code')`. However, it also has the following additional options:
 - `timeout`: a number of milliseconds to execute `code` before terminating
   execution. If execution is terminated, an [`Error`][] will be thrown.
 
+## Example: Run a Server within a VM
```
I think this should either not be a heading, or a 3rd level heading.

eljefedelrodeodeljefe replied:
I think we are using the heading style when it is an example, outside of a method, but rather a guide or such. E.g. here https://github.com/nodejs/node/blob/master/doc/api/readline.markdown#example-tiny-cli . If you'd rather want it belonging to `.runInThisContext()`, I can do it w/o pounds.

Having a look at for example `streams`, it would be 3rd level, but because it's having 2nd level topics like `Writable` etc. We are a little inconsistent about this I fear. :(

No strong opinions though.

jasnell commented on doc/api/vm.markdown:
```diff
@@ -301,6 +301,34 @@ e.g. `(0,eval)('code')`. However, it also has the following additional options:
+will need to apply additional objects to it. Say, you want to run a simple
+web server based on the `http` module, you can achieve this by doing something
+like this:
+
```
Looking at this again, I'd generally prefer staying away from "you" language... this can be reworded as:

```
The context of `.runInThisContext()` refers to the v8 context. The code passed 
to this VM context will have it's own isolated scope. To run a simple web server 
using the `http` module, for instance, the code passed to the context must either 
call `require('http')` on it's own, or have a reference to the `http` module passed 
to it. For instance:
```

eljefedelrodeodeljefe commented:
@jasnell agree. Changed it to what you provided.

Also added this issue, for potential harmonization of pronouns https://github.com/nodejs/docs/issues/87

Qard commented:
Not sure how I feel about documenting this in this way. Normally the vm functions are meant to provide a sandboxed context and this is doing the opposite. The problem I see here is that it's not immediately clear that require is not just a stateless function. Using require will stuff modules into the cache *globally*, and you could do some nefarious things, like mess with the contents of `require.cache`. If we *do* want to document this, I think we need to be *very* clear about the risks involved.

As a side note: I'd only really consider `eval()` to be an anti-pattern when used on user-supplied code. In a controlled environment, I'd consider it at worst to be a possible deoptimization trigger. It's not actually *that* bad, if you are careful.

eljefedelrodeodeljefe commented:
I'd wish for the API to be a little less confusing, but it is what is and at least it allows for some configuration. Also I wouldn't use `eval()` because you can't directly return from it. The use case I found for it is in a child_process - possibly - on a remote system. The layer of security then would be to share explicitly share resources between parent and child or not.

Documenting dangers would be fine of course. Maybe also hinting to the use case above.

jasnell commented:
@Qard @eljefedelrodeodeljefe ... any further thoughts on this one?

Qard commented:
It worries me a little, but I think I'm :+1: overall. LGTM.

Might be worth discussing more in a docs meeting, but I think we can add to it later rather than blocking.

TheAlphaNerd commented:
@eljefedelrodeodeljefe I was just about to land this but noticed the commit message is not so informative. Would you be able to update that? LGTM otherwise

eljefedelrodeodeljefe commented:
I have updated the commit message to reflect intention. Also I have added @Qard's remarks concerning the statefulness of require in this case in a note-section below the example. 

Also happy in case we need broader discussion

jasnell commented:
LGTM

thefourtheye commented on doc/api/vm.markdown:
```diff
@@ -301,6 +301,38 @@ e.g. `(0,eval)('code')`. However, it also has the following additional options:
 
+## Example: Run a Server within a VM
+
+The context of `.runInThisContext()` refers to the v8 context. The code passed
```
Minot nit. I think it is V8.

thefourtheye commented on doc/api/vm.markdown:
```diff
@@ -301,6 +301,38 @@ e.g. `(0,eval)('code')`. However, it also has the following additional options:
+## Example: Run a Server within a VM
+
+The context of `.runInThisContext()` refers to the v8 context. The code passed
+to this VM context will have it's own isolated scope. To run a simple web server
```
Isn't it `its`?

thefourtheye commented on doc/api/vm.markdown:
````diff
@@ -301,6 +301,38 @@ e.g. `(0,eval)('code')`. However, it also has the following additional options:
+to it. For instance:
+
+```js
+'use strict'
````
Missing semi-colons

eljefedelrodeodeljefe replied:
Which ones? After `'use strict'` is not necessary, since it works and activates strict mode.

thefourtheye replied:
In our repo every piece of code will have semicolons to mark the end of line. This example doesn't conform to that practice

thefourtheye commented on doc/api/vm.markdown:
````diff
@@ -301,6 +301,38 @@ e.g. `(0,eval)('code')`. However, it also has the following additional options:
+```   
+
+_Note: `require()` in the above case shares the state with context it is passed
+from. This might introduce risks when unknown code is executed._
````
Can we expand this a little bit? Personally, I would love to know the potential problems, if there is risk involved. Unknown code here means untrusted code or user supplied code?

addaleax replied:
I think it means untrusted code. But +1 on describing the potential problems. (The vm docs are a little vague for my taste with respect to that aspect in general…)

eljefedelrodeodeljefe replied:
As mentioned above, you likely need to run `vm` anyhow in separate process, that has non critical rights and contexts. `vm` is not constructed out-of-box to provide a infinitely safe sandbox, imo. I will add the process part and will add a subordinate clause that: you can mess with the calling threads objects and abuse its context.

addaleax commented on doc/api/vm.markdown:
```diff
@@ -301,6 +301,38 @@ e.g. `(0,eval)('code')`. However, it also has the following additional options:
+  })
+`
+
+vm.runInThisContext(code)(require)
```
Shouldn’t this be `vm.runInNewContext`? Otherwise it seems meaningless to me to point of security issues, because the whole set of current globals is available anyway.

eljefedelrodeodeljefe replied:
This will not work, because then you'd likely need to pass all objects you want to use including `console`. Also goal for this example is not security, it's running node code. For more security you'd want to run this in a separate process.

```console
evalmachine.<anonymous>:11
    console.log('Server running at http://127.0.0.1:8124/');
    ^
ReferenceError: console is not defined
    at evalmachine.<anonymous>:11:5
    at Object.<anonymous> (/Users/jefe/repos/node_rename_2/runinnewcontext.js:18:25)
    at Module._compile (module.js:413:34)
    at Object.Module._extensions..js (module.js:422:10)
    at Module.load (module.js:357:32)
    at Function.Module._load (module.js:314:12)
    at Function.Module.runMain (module.js:447:10)
    at startup (node.js:146:18)
    at node.js:404:3
```

eljefedelrodeodeljefe commented:
Addressed latest comments and force pushed

jasnell commented:
ping @thefourtheye

eljefedelrodeodeljefe commented:
rebased. Also tried to contact @thefourtheye concerning sign-off on slack. But I think he is just incredibly busy.

thefourtheye commented:
Really sorry about the delay. LGTM.

jasnell commented:
Awesome, ok, one final ping to @nodejs/documentation as a last call for comments.

benjamingr commented:
LGTM

Qard commented:
LGTM

jasnell commented:
Landed in 6815a3b. Fixed up the commit log a bit

TheAlphaNerd commented:
@jasnell / @eljefedelrodeodeljefe is this something we want for lts?

TheAlphaNerd commented:
ping @eljefedelrodeodeljefe @jasnell
